File: algorithms/code/leetcode/lc136_single_number/lc136_single_number.py
# https://leetcode.com/problems/single-number/

import unittest


class Solution:
    def singleNumber(self, nums: list[int]) -> int:
        result = 0
        for value in nums:
            result ^= value
        return result


class TestCode(unittest.TestCase):
    def test_1(self) -> None:
        self.assertEqual(1, Solution().singleNumber([1]))

    def test_1_2_1(self) -> None:
        self.assertEqual(2, Solution().singleNumber([1, 2, 1]))

File: algorithms/code/leetcode/lc136_single_number/lc136_single_number.rs
// https://leetcode.com/problems/single-number/
// #easy

pub struct Solution;

impl Solution {
    pub fn single_number(nums: Vec<i32>) -> i32 {
        let mut result = 0;
        for value in nums {
            result ^= value;
        }
        result
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_1() {
        assert_eq!(Solution::single_number(vec![1]), 1);
    }

    #[test]
    fn test_1_and_2_and_1() {
        assert_eq!(Solution::single_number(vec![1, 2, 1]), 2);
    }
}
